File: LeetCode75_LevelTwo/Implement_Trie_(Prefix_Tree)/Implement_Trie_(Prefix_Tree).py
# 208. Implement Trie (Prefix Tree) Python
class Trie(object):

    def __init__(self):
        self.trie = {}

    def insert(self, word):
        """
        :type word: str
        :rtype: None
        """
        t = self.trie
        for c in word:
            if c not in t: t[c] = {}
            t = t[c]
        t["-"] = True

    def search(self, word):
        """
        :type word: str
        :rtype: bool
        """
        t = self.trie
        for c in word:
            if c not in t: return False
            t = t[c]
        return "-" in t

    def startsWith(self, prefix):
        """
        :type prefix: str
        :rtype: bool
        """
        
        t = self.trie
        for c in prefix:
            if c not in t: return False
            t = t[c]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: LeetCode75_LevelTwo/Implement_Trie_(Prefix_Tree)/Implement_Trie_(Prefix_Tree).cc
// 208. Implement Trie (Prefix Tree) C++
#include <string>
using namespace std;

class TrieNode{
public:
    bool isEnd;
    TrieNode* children[26];
    TrieNode(){
        isEnd = false;
        for (int i = 0; i < 26; i++){
            children[i] = NULL;
        }
    }
};

class Trie {
public:
    TrieNode* root;
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        int l = word.size();
        TrieNode* parent = root;
        for (int i = 0; i < l; i++) {
            int val = word[i]-'a';
            if (parent->children[val] == NULL){
                parent->children[val] = new TrieNode();
            }
            parent = parent->children[val];
        }
        parent->isEnd = true;
    }
    
    bool search(string word) {
        int l = word.size();
        TrieNode* parent = root;
        for (int i = 0; i < l; i++) {
            int val = word[i]-'a';
            if (parent->children[val] == NULL){
                return false;
            }
            parent = parent->children[val];
        }
        return parent->isEnd;
    }
    
    bool startsWith(string prefix) {
        int l = prefix.size();
        TrieNode* parent = root;
        for (int i = 0; i < l; i++) {
            int val = prefix[i]-'a';
            if (parent->children[val] == NULL){
                return false;
            }
            parent = parent->children[val];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
